src/sources/kafka.rs

```diff
@@ -370,105 +411,559 @@ impl SourceConfig for KafkaSourceConfig {
+    /// the `finish_drain` method will return a Complete state, otherwise
+    /// a Consuming state.
+    shutdown: bool,
+}
+type OptionDeadline = OptionFuture<Pin<Box<Sleep>>>;
+enum ConsumerState {
+    Consuming(ConsumerStateInner<Consuming>),
+    Draining(ConsumerStateInner<Draining>),
+    Complete(ConsumerStateInner<Complete>),
+}
+impl Draining {
+    fn new(signal: SyncSender<()>, shutdown: bool) -> Self {
+        Self {
+            signal,
+            shutdown,
+            expect_drain: HashSet::new(),
+        }
     }
 
-    let mut stream = consumer.stream();
+    fn is_complete(&self) -> bool {
+        self.expect_drain.is_empty()
+    }
+}
 
-    loop {
-        tokio::select! {
-            biased;
-            _ = &mut shutdown => break,
-            entry = ack_stream.next() => if let Some((status, entry)) = entry {
-                if status == BatchStatus::Delivered {
-                    if let Err(error) =
-                        consumer.store_offset(&entry.topic, entry.partition, entry.offset)
-                    {
-                        emit!(KafkaOffsetUpdateError { error });
+impl<C> ConsumerStateInner<C> {
+    pub fn complete(self, _deadline: OptionDeadline) -> (OptionDeadline, ConsumerState) {
+        (
+            None.into(),
+            ConsumerState::Complete(ConsumerStateInner {
+                config: self.config,
+                decoder: self.decoder,
+                out: self.out,
+                log_namespace: self.log_namespace,
+                consumer_state: Complete,
+            }),
+        )
+    }
+}
+
+impl ConsumerStateInner<Consuming> {
+    const fn new(
+        config: KafkaSourceConfig,
+        decoder: Decoder,
+        out: SourceSender,
+        log_namespace: LogNamespace,
+    ) -> Self {
+        Self {
+            config,
+            decoder,
+            out,
+            log_namespace,
+            consumer_state: Consuming,
+        }
+    }
+
+    /// Spawn a task on the provided JoinSet to consume the kafka StreamPartitionQueue, and handle acknowledgements for the messages consumed
+    /// Returns a channel sender that can be used to signal that the consumer should stop and drain pending acknowledgements,
+    /// and an AbortHandle that can be used to forcefully end the task.
+    fn consume_partition(
+        &self,
+        join_set: &mut JoinSet<(TopicPartition, PartitionConsumerStatus)>,
+        tp: TopicPartition,
+        consumer: Arc<StreamConsumer<KafkaSourceContext>>,
+        p: StreamPartitionQueue<KafkaSourceContext>,
+        acknowledgements: bool,
+        exit_eof: bool,
+    ) -> (oneshot::Sender<()>, tokio::task::AbortHandle) {
+        // TODO there's probably a better way to pass the keys() config around than cloning the entire config object? But they are tied to the lifetime of the config object, so..this avoids lifetime messes for now
+        let conf = self.config.clone();
```

Comment:
Could you clone just the keys and use a reference to that? Maybe with `self.config.keys().cloned().collect::<Vec<_>>()` if needed?

Reply:
This didn't work, I think because the `Keys` struct here doesn't implement `Iterator` which would give us that `cloned()` method?
The best I've come up with is similar though: if we make the `Keys` struct own its members, instead of containing references, then each of the key paths from the config gets cloned for each consumed partition. That is a bit better than cloning the entire config object for each. I think it is possible to set up the lifetimes in such a way that these clones aren't needed, but I haven't been able to get my head around it so far.